crates/telio-starcast/src/nat.rs

```diff
@@ -0,0 +1,648 @@
+    fn translate_incomming(&mut self, packet: &mut [u8]) -> Result<(), Error>;
+}
+
+type Mapper = fn(src_addr: SocketAddr) -> u16;
+
+fn default_mapper(src_addr: SocketAddr) -> u16 {
+    let mut hasher = DefaultHasher::new();
+    match src_addr.ip() {
+        IpAddr::V4(src_ip) => Hash::hash_slice(&src_ip.octets(), &mut hasher),
+        IpAddr::V6(src_ip) => Hash::hash_slice(&src_ip.octets(), &mut hasher),
+    };
+    src_addr.port().hash(&mut hasher);
+    hasher.finish() as u16 // TODO: think if we want to use the full 16 bits or maybe some range e.g. >=1024
+}
+
+pub struct StarcastNat {
+    mappings: LruCache<u16, SocketAddr>,
+    vpeer_ipv4_addr: Ipv4Addr,
+    vpeer_ipv6_addr: Ipv6Addr,
+    mapper: Mapper,
+}
+
+impl StarcastNat {
+    /// Create new starcast nat with default capcity of 2048 entries and 2 minute timeout
+    /// 2min default is taken from (https://datatracker.ietf.org/doc/html/rfc4787#section-4.3)
+    pub fn new(vpeer_ipv4: Ipv4Addr, vpeer_ipv6: Ipv6Addr) -> Self {
+        const LRU_CAPACITY: usize = 2048;
+        const LRU_TIMEOUT: Duration = Duration::from_secs(120);
+        Self::new_custom(
+            vpeer_ipv4,
+            vpeer_ipv6,
+            LRU_CAPACITY,
+            LRU_TIMEOUT,
+            default_mapper,
+        )
+    }
+
+    /// Create new starcast nat with custom capacity and timeout
+    fn new_custom(
+        vpeer_ipv4_addr: Ipv4Addr,
+        vpeer_ipv6_addr: Ipv6Addr,
+        capacity: usize,
+        timeout: Duration,
+        mapper: Mapper,
+    ) -> Self {
+        Self {
+            mappings: LruCache::new(timeout, capacity),
+            vpeer_ipv4_addr,
+            vpeer_ipv6_addr,
+            mapper,
+        }
+    }
+
+    fn translate_incomming_internal<'a, P: MutableIpPacket<'a>>(
+        &mut self,
+        packet: &'a mut [u8],
+    ) -> Result<(), Error> {
+        let mut ip_packet = P::new(packet).ok_or(Error::PacketTooShort)?;
+        if ip_packet.get_next_level_protocol() != IpNextHeaderProtocols::Udp {
+            return Err(Error::UnexpectedTransportProtocol);
+        }
+
+        let src_ip = ip_packet.get_source();
+        let dst_ip = ip_packet.get_destination();
+
+        match src_ip {
+            IpAddr::V4(_) => {
+                ip_packet.set_source(IpAddr::V4(self.vpeer_ipv4_addr));
+                ip_packet.set_checksum(ipv4::checksum(&ip_packet.to_immutable_ipv4()));
+            }
+            IpAddr::V6(_) => {
+                ip_packet.set_source(IpAddr::V6(self.vpeer_ipv6_addr));
+            }
+        };
+
+        let mut udp_packet =
+            MutableUdpPacket::new(ip_packet.payload_mut()).ok_or(Error::PacketTooShort)?;
+        let src_port = udp_packet.get_source();
+        let src_addr = SocketAddr::new(src_ip, src_port);
+        let mut nat_port = (self.mapper)(src_addr);
+
+        while let Some(&current_mapping) = self.mappings.peek(&nat_port) {
```

Comment:
This introduces a posibility of infinite loop by a malicious user, somebody could span a a lot of quick packets of multicast, and make this part loop for longer and longer periods.
Would be nice to add some kind of protection.

Reply:
I thought about it but I don't know if this is really a big deal. And what could be a possible protection? Remember that those are the packets coming from the other peer via transport socket. You could in theory limit this peer to some number of concurrent entries so that the other peers are not affected by some malicious software running on that peer, but other than that I don't know. Any ideas?

Reply:
So in reality my main fear would be a compleate fill up of allowed ports, as now its possible (probably hard, and lru should mitigate greatly), but a simple check that we dont execeed u16::MAX would suffice for me.
or we could even hallf it so in avergage case you would need only two tries to find empty place.

Further improving we can leave for later on

Reply:
As now, because you are using peek, if at any point somebody manages to spam all 2^16 port's, you fall into loop that would never exit.

Reply:
I've limited the attempts to some hardcoded value to fix the infinite loop issue. That's the thing Linux kernel does in their nat implementation (as a last resort).